contrib/fparser/extrasrc/Makefile.am

```diff
@@ -15,3 +15,10 @@ fp_opcode_add.inc: $(srcdir)/fp_opcode_add.inc.release
 	$(AM_V_GEN)$(LN_S) -f $< $@
 
 endif
+
+# We need to install some additional headers to permit
+# developers to derive from FParser to implement evaluation
+# on custom types.
+includedir = $(prefix)/include/libmesh/extrasrc
```

Comment:
fparser doesn't really follow the pattern of our other contrib libraries that have multiple headers, e.g. timpi, poly2tri, metaphysicl, which get installed in `include/libmesh/timpi`, etc. I don't really like the idea of having an "extrasrc" directory at the same level as these other libraries, with the other fparser headers being at include/libmesh/fparser.hh and include/libmesh/fparser_ad.hh... Given that we are shaking things up anyway, I think this would be a good time to make an `include/libmesh/fparser` top level include directory. I think we could also make it backwards compatible by setting the include search paths correctly in the Makefile.

Reply:
There's no `include/libmesh/timpi` - we install to `include/timpi`.  So `include/fparser` would be the way to go there ... except that, for `include/timpi` and the like, they all have one-way dependencies - they can be installed separately, they can be compiled against and linked to even if the rest of libMesh isn't there, they don't require anything from the libMesh directory, etc.  But as long as our fparser fork is still demanding `libmesh_config.h` rather than being operable as a standalone fork, I think `include/libmesh/fparser` might be a reasonable way to organize it.

Reply:
There's `LIBMESH_USING_THREADS`, `LIBMESH_HAVE_FPARSER_JIT`, and `LIBMESH_HAVE_CXX11_ERF` which we use in FParser.

Reply:
> I think we could also make it backwards compatible by setting the include search paths correctly in the Makefile.

Ideally we would want `libmesh-config` spit out the additional `fparser` include path. I have no idea where to make that change.

Reply:
Actually.. sigh, not that easy. I see where the path is set in `contrib/bin/libmesh-config.in`, but we do not add `libmesh` to that path (unless in legacy include mode). So the include statements are

currently:
```
#inlcude "libmesh/fparser.hh"
```

proposed:
```
#inlcude "libmesh/fparser/fparser.hh"
```
There is not additional path we can add that would make that change backwards compatible.

Reply:
I see... the issue is that fparser doesn't have it's own `./configure`, so it's relying on a bunch of flags determined by libmesh's configure... other than that, the coupling is basically a "one-way" coupling like the rest of the contrib libraries.

Reply:
> There is not additional path we can add that would make that change backwards compatible.

OK, I guess the backwards compatibility layer will have to be us making a symbolic link from the existing
`include/libmesh/fparser.hh` -> `include/libmesh/fparser/fparser.hh`
?

Then we'd give people time to update their codes from:
```
#include "libmesh/fparser.hh"
```
to 
```
#include "libmesh/fparser/fparser.hh"
```
and eventually remove the links after some deprecation period.

Reply:
Uhm, and how do we make that symlink?

Reply:
That's a good question. We have this script, `include/libmesh/rebuild_makefile.sh` that you normally have to run whenever you add/remove files. It generates the `include/libmesh/Makefile.am` file. If you look at that Makefile.am, you should see that we already do some special handling of the fparser headers... I think maybe here is where we could set up additional links, but it would definitely take me some trial and error to figure this out. 
```
#
# contrib rules
if LIBMESH_ENABLE_FPARSER

fparser.hh: $(top_srcdir)/contrib/fparser/fparser.hh
	$(AM_V_GEN)rm -f $@ && $(LN_S) -f $< $@

fparser_ad.hh: $(top_srcdir)/contrib/fparser/fparser_ad.hh
	$(AM_V_GEN)rm -f $@ && $(LN_S) -f $< $@

  BUILT_SOURCES  += fparser.hh fparser_ad.hh
  DISTCLEANFILES += fparser.hh fparser_ad.hh

endif
```

Reply:
We need to set up the links when we install though not before. What should have been a small PR is now turning into an autotools nightmare for me :-(

Reply:
OK, well since @roystgnr is OK with the change as-is, I guess it's fine with me. I am wondering though if you have actually tested including this new header in some way at the MOOSE level? We don't have any unit tests or examples that use the new fparser header, so just passing the CI testing here doesn't tell us much.

Reply:
> I am wondering though if you have actually tested including this new header in some way at the MOOSE level?

I have a MOOSE app for an LDRD in which I've been using this libmesh branch for almost a week to implement a GPU accelerated function parser.